src/librbd/api/Migrate.cc

```diff
@@ -0,0 +1,991 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#include "librbd/api/Migrate.h"
+#include "include/rados/librados.hpp"
+#include "common/dout.h"
+#include "common/errno.h"
+#include "cls/rbd/cls_rbd_client.h"
+#include "librbd/ExclusiveLock.h"
+#include "librbd/ImageCtx.h"
+#include "librbd/ImageState.h"
+#include "librbd/Operations.h"
+#include "librbd/Utils.h"
+#include "librbd/deep_copy/SetHeadRequest.h"
+#include "librbd/deep_copy/SnapshotCopyRequest.h"
+#include "librbd/image/CreateRequest.h"
+#include "librbd/image/RemoveRequest.h"
+#include "librbd/internal.h"
+#include "librbd/io/ImageRequestWQ.h"
+#include "librbd/mirror/DisableRequest.h"
+#include "librbd/mirror/EnableRequest.h"
+
+#define dout_subsys ceph_subsys_rbd
+#undef dout_prefix
+#define dout_prefix *_dout << "librbd::Migrate: " << __func__ << ": "
+
+namespace librbd {
+namespace api {
+
+namespace {
+
+int trash_search(librados::IoCtx &io_ctx, rbd_trash_image_source_t source,
+                 const std::string &image_name, std::string *image_id) {
+  std::vector<trash_image_info_t> entries;
+
+  int r = trash_list(io_ctx, entries);
+  if (r < 0) {
+    return r;
+  }
+
+  for (auto &entry : entries) {
+    if (entry.source == source && entry.name == image_name) {
+      *image_id = entry.id;
+      return 0;
+    }
+  }
+
+  return -ENOENT;
+}
+
+} // anonymous namespace
+
+template <typename I>
+int Migrate<I>::migrate(I *ictx, librados::IoCtx& dst_io_ctx,
+                        const char *dstname_, ImageOptions& opts,
+                        ProgressContext &prog_ctx) {
+  CephContext* cct = ictx->cct;
+
+  std::string dstname = (dstname_ != nullptr && *dstname_ != '\0') ?
```

Comment:
Nit: ```!dstname_->empty()```

Reply:
`dstname_` is `const char *`. It could be `const string&`, I just followed the convention that it looks we use `const char *` in API functions.